### ISSUE TITLE
Non-reference resource joins

### ISSUE THREAD

**Issue by johngrimes** [aehrc/pathling#338]
Sometimes within the FHIR model, resources (often definitional) are referenced by other resources without using the regular resource reference mechanism. An example of this is `QuestionnaireResponse.questionnaire` referencing `Questionnaire.url`.

We need some other mechanism that perhaps allows the user to make joins between resources using arbitrary criteria.

**Comment by johngrimes** [comment]
I'm reading the spec a bit more closely, and it looks like we could just add special support for `canonical` to the existing `resolve` function.

https://hl7.org/fhir/R4/fhirpath.html#functions

So it might look something like this:

```
questionnaire.resolve().name
```

/cc @developerdu

**Comment by johngrimes** [comment]
We could take the same approach as Ontoserver, anointing a couple of versioning schemes and leaving the behaviour of others undefined. I think it supports [SemVer](https://semver.org/) and `YYYYMMDD`.

**Comment by developerdu** [comment]
Agreed, questionnaire.resolve().name was what I was hoping would work.  Good issue name change too, I searched for "Canonical" to try and find an existing issue but didn't hit on this one.

**Comment by johngrimes** [comment]
Here's the section of the spec that talks about canonical URLs: https://www.hl7.org/fhir/references.html#canonical

**Comment by johngrimes** [comment]
Within the context of SQL on FHIR, this feature now moves to become a capability of the `getReferenceKey` function.
